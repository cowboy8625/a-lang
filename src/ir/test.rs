// use crate::semantic_analysis::{SemanticAnalysisVisitor, SymbolTableBuilder};
macro_rules! snapshot {
    ($name:tt, $path:tt) => {
        #[test]
        fn $name() {
            use super::*;
            use $crate::lexer::lex;
            use $crate::parse::parse;
            let contents = include_str!($path);
            let tokens = lex(contents).unwrap();
            let ast = parse(tokens).unwrap();
            // let mut symbol_table_builder = SymbolTableBuilder::default();
            // symbol_table_builder.visit(&ast);
            // let symbol_table = symbol_table_builder.build();
            let ir_code = code_gen(ast).unwrap();
            let result = ir_code
                .iter()
                .map(ToString::to_string)
                // .map(|i| format!("{i:#?}\n"))
                .collect::<String>();
            let mut settings = insta::Settings::clone_current();
            settings.set_snapshot_path("testdata/output/");
            settings.bind(|| {
                insta::assert_snapshot!(result);
            });
        }
    };
}

snapshot!(binary, "../../snapshots/binary.a");
snapshot!(ifelse, "../../snapshots/ifelse.a");
snapshot!(max, "../../snapshots/max.a");
