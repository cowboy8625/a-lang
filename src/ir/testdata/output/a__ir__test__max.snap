---
source: src/ir/test.rs
expression: result
---
function max(%0: u64, %1: u64) {
    enter
    %2 = %0 > %1
    if %2 goto .L0
    return %0
    goto .exit
.L0:
    return %1
    goto .exit
.exit:
    leave
}function main() {
    enter
    load %1 1
    load %2 2
    call max(%1, %2) -> %0
    return
    goto .exit
.exit:
    leave
}
